core/src/main/java/io/grpc/internal/DnsNameResolver.java

```diff
@@ -226,4 +233,146 @@ public final synchronized void shutdown() {
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Common interface between the delegate resolvers used by DnsNameResolver.
+   */
+  @VisibleForTesting
+  abstract static class DelegateResolver {
+    abstract ResolutionResults resolve(String host) throws Exception;
+  }
+
+  /**
+   * Describes the results from a DNS query.
+   */
+  @VisibleForTesting
+  static final class ResolutionResults {
+    final List<InetAddress> addresses;
+    final List<String> txtRecords;
+
+    ResolutionResults(List<InetAddress> addresses, List<String> txtRecords) {
+      this.addresses = Collections.unmodifiableList(checkNotNull(addresses, "addresses"));
+      this.txtRecords = Collections.unmodifiableList(checkNotNull(txtRecords, "txtRecords"));
+    }
+  }
+
+  /**
+   * A composite DNS resolver that uses both the JDK and JNDI resolvers as delegate.  It is
+   * expected that two DNS queries will be executed, with the second one being from JNDI.
+   */
+  @VisibleForTesting
+  static final class CompositeResolver extends DelegateResolver {
+
+    private final DelegateResolver jdkResovler;
+    private final DelegateResolver jndiResovler;
+
+    CompositeResolver(DelegateResolver jdkResovler, DelegateResolver jndiResovler) {
+      this.jdkResovler = jdkResovler;
+      this.jndiResovler = jndiResovler;
+    }
+
+    @Override
+    ResolutionResults resolve(String host) throws Exception {
+      ResolutionResults results = jdkResovler.resolve(host);
+      logger.info(results.addresses.toString());
+      List<InetAddress> addresses = results.addresses;
+      ResolutionResults jdniResults = jndiResovler.resolve(host);
+      List<String> txtRecords = jdniResults.txtRecords;
+
+      return new ResolutionResults(addresses, txtRecords);
+    }
+  }
+
+  /**
+   * The default name resolver provided with the JDK.  This is unable to lookup TXT records, but
+   * provides address ordering sorted according to RFC 3484.  This is true on OpenJDK, because it
+   * in turn calls into libc which sorts addresses in order of reachability.
+   */
+  @VisibleForTesting
+  static final class JdkResolver extends DelegateResolver {
+
+    @Override
+    ResolutionResults resolve(String host) throws Exception {
+      return new ResolutionResults(
+          Arrays.asList(InetAddress.getAllByName(host)),
+          Collections.<String>emptyList());
+    }
+  }
+
+  /**
+   * A resolver that uses the JNDI interface.  This class is capable of looking up both addresses
+   * and text records, but does not provide ordering guarantees.  When the {@link JdkResolver} is
+   * not available, this can be used as a fallback.  Some results are better than no results.
+   */
+  @VisibleForTesting
+  static final class JndiResolver extends DelegateResolver {
+
+    private static final String[] rrTypes = new String[]{"TXT"};
+
+    @Override
+    ResolutionResults resolve(String host) throws NamingException {
```

Comment:
Do we want to catch any exceptions from this, so that a failure to get TXT doesn't hose everything?

We may want to discuss with others what should happen if we get A/AAAA but have I/O errors durring TXT.